IPNI_FEDERATION.md

```diff
@@ -0,0 +1,417 @@
+its own. Discrepancies are flagged based on a set of predefined conflict resolution rules. For instance, if two
+instances have added different data for the same key, the system might prioritize data from the instance with a more
+recent vector clock timestamp.
+
+Reconciliation is not just about resolving conflicts. It's also about merging updates. If an instance discovers newer
+data from a snapshot, it updates its index accordingly. Through continuous reconciliation processes, the protocol
+ensures that all IPNI instances trend towards eventual consistency.
+
+### Conflict Resolution
+
+In the vast landscape of the Inter-Planetary Network Indexer (IPNI), advertisement chains serve as the core mechanism
+for content propagation. As multiple IPNI instances exist across the globe, the potential for discrepancies in the state
+of these chains is undeniable. It becomes essential, therefore, to devise strategies that can reconcile these
+discrepancies effectively.
+
+#### Case 1: A provider is unknown by another IPNI instance
+
+**Scenario:**  
+Instance A is aware of Provider P1 and its advertisement chain, but Instance B is unaware of Provider P1 entirely.
+
+**Resolution Strategy:**  
+When instances exchange information, and Instance B identifies a missing provider, it should:
+
+1. Request the complete advertisement chain of Provider P1 from Instance A.
+2. Incorporate Provider P1's advertisement chain into its local index.
+3. Broadcast this new knowledge to other connected IPNI instances to ensure widespread awareness and data consistency.
+
+#### Case 2: A provider is known by both with a consistent head advertisement
+
+**Scenario:**  
+Both Instance A and Instance B are aware of Provider P1 and have the same head (latest) advertisement.
+
+**Resolution Strategy:**  
+This represents the ideal state of eventual consistency. In this scenario:
+
+1. Both instances recognize that they have the same version of the advertisement chain for Provider P1.
+2. No reconciliation action is required.
+3. The instances can continue to propagate other advertisement chains or updates, knowing that, for Provider P1, they
+   are in sync.
+
+#### Case 3: A provider is known by both but with different head advertisements
+
+**Scenario:**  
+Instance A and Instance B both know Provider P1, but they have different latest advertisements (heads) for P1's chain.
+
+**Resolution Strategy:**  
+This represents a divergence in the advertisement chain, and careful reconciliation is required:
+
+1. Instance A and B determine which advertisement is the later one in the chain of advertisements provided by P1.
+2. They converge on the advertisement has been published the latest, i.e. has the shorter distance to P1's current 
+   advertisement head.
+
+While traversing advertisement chains, instances should prefer pulling ad chain from eachother's mirror instead of going
+directly to the provider where possible.
```

Comment:
should we include the height of each head in snapshots, so that the understanding of which one is 'latest' is easier to resolve?

Reply:
We could but would that number be hearsay? I think height needs to be added to ad by provider, and signed instead but that's a much bigger change

Aside from trust, we also want to encourage indexers to fetch ad chains from each other's mirrors to replicate the chain. 

Thoughts?

Reply:
one of the two parties should know the other parties head already since it will be older. A 'latest' provides a cheap hint for who is supposed to initiate the sync, but an un-trusting indexer could ignore it and still follow the protocol correctly.

Reply:
it seems like an easy enough thing

Reply:
Also captured https://github.com/ipni/specs/issues/28

Reply:
This seems unfair to the faster indexer, because all the other indexers will query the mirror of the fastest. Maybe we need a strategy that strongly encourages other indexers to add content from others' mirrors to their own.  

For example, an indexer cannot pull advertisement data directly from another mirror, but instead must sync some portion of the mirror and then get the advertisements from the local mirror after syncing.

The fast source indexer may then redirect others, for a short time, to the destination indexer for the synced content. Or, maybe there is some check that if one indexer copies advertisement data from another indexer's mirror, the indexer that received the data subsequently serves it from its mirror.